README.md
# Dopple

HTTP proxy to Unix Socket based JSON-RPC servers

## Installation

```sh
pip install dopple
```

## Use

```sh
dopple ~/.ethereum/geth.ipc http://127.0.0.1:8545
```

Alternatively, `dopple` can be invoked as a pure script without installation:

```sh
./dopple/dopple.py ~/.ethereum/geth.ipc http://127.0.0.1:8545
```

These values above are the default ones too. If they match your current configuration, they can be ommitted.

Check out `--help` option for more information.

Read more in the [documentation on ReadTheDocs](https://dopple.readthedocs.io/). [View the change log](https://dopple.readthedocs.io/en/latest/releases.html).

## License

Apache-2.0

## Developer Setup

If you would like to hack on dopple, please check out the [Snake Charmers
Tactical Manual](https://github.com/ethereum/snake-charmers-tactical-manual)
for information on how we do:

- Testing
- Pull Requests
- Code Style
- Documentation

### Development Environment Setup

You can set up your dev environment with:

```sh
git clone git@github.com:ethereum/dopple.git
cd dopple
virtualenv -p python3 venv
. venv/bin/activate
pip install -e .[dev]
```

### Testing Setup

During development, you might like to have tests run on every file save.

Show flake8 errors on file change:

```sh
# Test flake8
when-changed -v -s -r -1 dopple/ tests/ -c "clear; flake8 dopple tests && echo 'flake8 success' || echo 'error'"
```

Run multi-process tests in one command, but without color:

```sh
# in the project root:
pytest --numprocesses=4 --looponfail --maxfail=1
# the same thing, succinctly:
pytest -n 4 -f --maxfail=1
```

Run in one thread, with color and desktop notifications:

```sh
cd venv
ptw --onfail "notify-send -t 5000 'Test failure ⚠⚠⚠⚠⚠' 'python 3 test on dopple failed'" ../tests ../dopple
```

### Release setup

For Debian-like systems:
```
apt install pandoc
```

To release a new version:

```sh
make release bump=$$VERSION_PART_TO_BUMP$$
```

#### How to bumpversion

The version format for this repo is `{major}.{minor}.{patch}` for stable, and
`{major}.{minor}.{patch}-{stage}.{devnum}` for unstable (`stage` can be alpha or beta).

To issue the next version in line, specify which part to bump,
like `make release bump=minor` or `make release bump=devnum`. This is typically done from the
master branch, except when releasing a beta (in which case the beta is released from master,
and the previous stable branch is released from said branch).

If you are in a beta version, `make release bump=stage` will switch to a stable.

To issue an unstable version when the current version is stable, specify the
new version explicitly, like `make release bump="--new-version 4.0.0-alpha.1 devnum"`


docs/release_notes.rst
Release Notes
=============

.. towncrier release notes start

v0.4.0-alpha.1
--------------

- Launched repository, claimed names for pip, RTD, github, etc


tests/core/test_import.py


def test_import():
    import dopple  # noqa: F401


scripts/jsonrpcproxy.py
#!/usr/bin/env python3

# Copyright 2017 The Dopple Authors.
# Licensed under the Apache License, Version 2.0. See the LICENSE file.

"""
JSON-RPC Proxy

This Python script provides HTTP proxy to Unix Socket based JSON-RPC servers.
Check out --help option for more information.

Build with cython:

cython rpcproxy.py --embed
gcc -O3 -I /usr/include/python3.5m -o rpcproxy rpcproxy.c \
-Wl,-Bstatic -lpython3.5m -lz -lexpat -lutil -Wl,-Bdynamic -lpthread -ldl -lm

"""

from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter
from http.server import HTTPServer, BaseHTTPRequestHandler
from os import path
from urllib.parse import urlparse

import errno
import socket
import sys
import time
import threading

if sys.platform == 'win32':
    import win32file
    import pywintypes


VERSION = '0.2'
BUFSIZE = 32
DELIMITER = ord('\n')
BACKEND_CONNECTION_TIMEOUT=30.0
INFO = """JSON-RPC Proxy

Version:  {version}
Proxy:    {proxy_url}
Backend:  {backend_url} (connected: {connected})
"""


class BackendError(Exception):
    pass


class UnixSocketConnector(object):
    """Unix Domain Socket connector. Connects to socket lazily."""

    def __init__(self, socket_path):
        self._socket_path = socket_path
        self._socket = None

    @staticmethod
    def _get_error_message(os_error_number):
        if os_error_number == errno.ENOENT:
            return "Unix Domain Socket '{}' does not exist"
        if os_error_number == errno.ECONNREFUSED:
            return "Connection to '{}' refused"
        return "Unknown error when connecting to '{}'"

    def socket(self):
        """Returns connected socket."""
        if self._socket is None:
            try:
                s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
                s.connect(self._socket_path)
                s.settimeout(1)
                # Assign last, to keep it None in case of exception.
                self._socket = s
            except OSError as ex:
                msg = self._get_error_message(ex.errno)
                err = BackendError(msg.format(self._socket_path))
                raise err from ex
        return self._socket

    def close(self):
        if self._socket is not None:
            self._socket.shutdown(socket.SHUT_RDWR)
            self._socket.close()
            self._socket = None

    def is_connected(self):
        return self._socket is not None

    def check_connection(self, timeout):
        SLEEPTIME = 0.1
        wait_time = 0.0
        last_exception = None
        while True:
            try:
                if self.socket():
                    break
            except BackendError as ex:
                last_exception = ex  # Ignore backed errors for some time.

            time.sleep(SLEEPTIME)
            wait_time += SLEEPTIME
            if wait_time > timeout:
                raise last_exception if last_exception else TimeoutError

    def recv(self, max_length):
        return self.socket().recv(max_length)

    def sendall(self, data):
        try:
            return self.socket().sendall(data)
        except OSError as ex:
            if ex.errno == errno.EPIPE:
                # The connection was terminated by the backend. Try reconnect.
                self.close()
                return self.socket().sendall(data)
            else:
                raise


class NamedPipeConnector(object):
    """Windows named pipe simulating socket."""

    def __init__(self, ipc_path):
        try:
            self.handle = win32file.CreateFile(
                ipc_path, win32file.GENERIC_READ | win32file.GENERIC_WRITE,
                0, None, win32file.OPEN_EXISTING, 0, None)
        except pywintypes.error as err:
            raise IOError(err)

    def is_connected(self):
        return True

    def check_connection(self, timeout):
        pass

    def recv(self, max_length):
        (err, data) = win32file.ReadFile(self.handle, max_length)
        if err:
            raise IOError(err)
        return data

    def sendall(self, data):
        return win32file.WriteFile(self.handle, data)

    def close(self):
        self.handle.close()


def get_ipc_connector(ipc_path):
    if sys.platform == 'win32':
        return NamedPipeConnector(ipc_path)
    return UnixSocketConnector(ipc_path)


class HTTPRequestHandler(BaseHTTPRequestHandler):

    def do_GET(self):
        if self.path != '/':
            self.send_response(404)
            self.end_headers()
            return

        self.send_response(200)
        self.send_header("Content-type", "text/plain")
        self.addCORS()
        self.end_headers()
        backend_url = 'unix:' + self.server.backend_address
        proxy_url = '{}:{}'.format(self.server.server_name,
                                   self.server.server_port)
        info = INFO.format(version=VERSION, proxy_url=proxy_url,
                           backend_url=backend_url,
                           connected=self.server.conn.is_connected())
        self.wfile.write(info.encode('utf-8'))

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Content-type", "text/plain")
        self.addCORS()
        self.end_headers()

    def do_POST(self):
        request_length = int(self.headers['Content-Length'])
        request_content = self.rfile.read(request_length)
        # self.log_message("Headers:  {}".format(self.headers))
        # self.log_message("Request:  {}".format(request_content))

        try:
            response_content = self.server.process(request_content)
            # self.log_message("Response: {}".format(response_content))

            self.send_response(200)
            self.send_header("Content-type", "application/json")
            self.send_header("Content-length", len(response_content))
            self.addCORS()
            self.end_headers()
            self.wfile.write(response_content)
        except BackendError as err:
            self.send_response(502)
            error_msg = str(err).encode('utf-8')
            # TODO: Send as JSON-RPC response
            self.send_header("Content-type", "text/plain")
            self.send_header("Content-length", len(error_msg))
            self.end_headers()
            self.wfile.write(error_msg)
            self.log_message("Backend Error: {}".format(err))

        # TODO: Handle other exceptions as error 500.

    def addCORS(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "content-type")


class Proxy(HTTPServer):

    def __init__(self, proxy_url, backend_path):
        self.proxy_url = proxy_url
        url = urlparse(proxy_url)
        assert url.scheme == 'http'
        proxy_address = url.hostname, url.port

        super(Proxy, self).__init__(proxy_address, HTTPRequestHandler)

        self.backend_address = path.expanduser(backend_path)

    def process(self, request):
        self.conn.sendall(request)

        response = b''
        while True:
            r = self.conn.recv(BUFSIZE)
            if not r:
                break
            if r[-1] == DELIMITER:
                response += r[:-1]
                break
            response += r

        return response

    def run(self):
        self.conn = get_ipc_connector(self.backend_address)
        self.conn.check_connection(timeout=BACKEND_CONNECTION_TIMEOUT)

        print("JSON-RPC HTTP Proxy: {} -> {}".format(
            self.backend_address, self.proxy_url), file=sys.stderr, flush=True)
        self.serve_forever()


if sys.platform == 'win32':
    DEFAULT_BACKEND_PATH = r'\\.\pipe\geth.ipc'
    BACKEND_PATH_HELP = "Named Pipe of a backend RPC server"
else:
    DEFAULT_BACKEND_PATH = '~/.ethereum/geth.ipc'
    BACKEND_PATH_HELP = "Unix Socket of a backend RPC server"

DEFAULT_PROXY_URL = 'http://127.0.0.1:8545'
PROXY_URL_HELP = "URL for this proxy server"


def parse_args():
    parser = ArgumentParser(
        description='HTTP Proxy for JSON-RPC servers',
        formatter_class=ArgumentDefaultsHelpFormatter
    )

    parser.add_argument('backend_path', nargs='?',
                        default=DEFAULT_BACKEND_PATH,
                        help=BACKEND_PATH_HELP)
    parser.add_argument('proxy_url', nargs='?',
                        default=DEFAULT_PROXY_URL,
                        help=PROXY_URL_HELP)
    return parser.parse_args()


def run(proxy_url=DEFAULT_PROXY_URL, backend_path=DEFAULT_BACKEND_PATH):
    proxy = Proxy(proxy_url, backend_path)
    try:
        proxy.run()
    except KeyboardInterrupt:
        proxy.shutdown()


def run_daemon(proxy_url=DEFAULT_PROXY_URL, backend_path=DEFAULT_BACKEND_PATH):
    proxy = Proxy(proxy_url, backend_path)
    th = threading.Thread(name='jsonrpcproxy', target=proxy.run)
    th.daemon = True
    th.start()
    return proxy


if __name__ == '__main__':
    args = parse_args()
    run(args.proxy_url, args.backend_path)



dopple/dopple.py
#!/usr/bin/env python3

# Copyright 2017 The Dopple Authors.
# Licensed under the Apache License, Version 2.0. See the LICENSE file.

"""
Dopple JSON-RPC Proxy

This Python script provides HTTP proxy to Unix Socket based JSON-RPC servers.
Check out --help option for more information.

Build with cython:

cython dopple.py --embed
gcc -O3 -I /usr/include/python3.5m -o dopple dopple.c \
-Wl,-Bstatic -lpython3.5m -lz -lexpat -lutil -Wl,-Bdynamic -lpthread -ldl -lm

"""

from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter
from http.server import HTTPServer, BaseHTTPRequestHandler
from os import path
from urllib.parse import urlparse

import errno
import pkg_resources
import socket
import sys
import time
import threading

if sys.platform == 'win32':
    import win32file
    import pywintypes

try:
    VERSION = pkg_resources.get_distribution("dopple").version
except pkg_resources.DistributionNotFound:
    VERSION = 'unknown'

BUFSIZE = 32
DELIMITER = ord('\n')
BACKEND_CONNECTION_TIMEOUT=30.0
INFO = """Dopple JSON-RPC Proxy

Version:  {version}
Proxy:    {proxy_url}
Backend:  {backend_url} (connected: {connected})
"""


class BackendError(Exception):
    pass


class UnixSocketConnector(object):
    """Unix Domain Socket connector. Connects to socket lazily."""

    def __init__(self, socket_path):
        self._socket_path = socket_path
        self._socket = None

    @staticmethod
    def _get_error_message(os_error_number):
        if os_error_number == errno.ENOENT:
            return "Unix Domain Socket '{}' does not exist"
        if os_error_number == errno.ECONNREFUSED:
            return "Connection to '{}' refused"
        return "Unknown error when connecting to '{}'"

    def socket(self):
        """Returns connected socket."""
        if self._socket is None:
            try:
                s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
                s.connect(self._socket_path)
                s.settimeout(1)
                # Assign last, to keep it None in case of exception.
                self._socket = s
            except OSError as ex:
                msg = self._get_error_message(ex.errno)
                err = BackendError(msg.format(self._socket_path))
                raise err from ex
        return self._socket

    def close(self):
        if self._socket is not None:
            self._socket.shutdown(socket.SHUT_RDWR)
            self._socket.close()
            self._socket = None

    def is_connected(self):
        return self._socket is not None

    def check_connection(self, timeout):
        SLEEPTIME = 0.1
        wait_time = 0.0
        last_exception = None
        while True:
            try:
                if self.socket():
                    break
            except BackendError as ex:
                last_exception = ex  # Ignore backed errors for some time.

            time.sleep(SLEEPTIME)
            wait_time += SLEEPTIME
            if wait_time > timeout:
                raise last_exception if last_exception else TimeoutError

    def recv(self, max_length):
        return self.socket().recv(max_length)

    def sendall(self, data):
        try:
            return self.socket().sendall(data)
        except OSError as ex:
            if ex.errno == errno.EPIPE:
                # The connection was terminated by the backend. Try reconnect.
                self.close()
                return self.socket().sendall(data)
            else:
                raise


class NamedPipeConnector(object):
    """Windows named pipe simulating socket."""

    def __init__(self, ipc_path):
        try:
            self.handle = win32file.CreateFile(
                ipc_path, win32file.GENERIC_READ | win32file.GENERIC_WRITE,
                0, None, win32file.OPEN_EXISTING, 0, None)
        except pywintypes.error as err:
            raise IOError(err)

    def is_connected(self):
        return True

    def check_connection(self, timeout):
        pass

    def recv(self, max_length):
        (err, data) = win32file.ReadFile(self.handle, max_length)
        if err:
            raise IOError(err)
        return data

    def sendall(self, data):
        return win32file.WriteFile(self.handle, data)

    def close(self):
        self.handle.close()


def get_ipc_connector(ipc_path):
    if sys.platform == 'win32':
        return NamedPipeConnector(ipc_path)
    return UnixSocketConnector(ipc_path)


class HTTPRequestHandler(BaseHTTPRequestHandler):

    def do_GET(self):
        if self.path != '/':
            self.send_response(404)
            self.end_headers()
            return

        self.send_response(200)
        self.send_header("Content-type", "text/plain")
        self.addCORS()
        self.end_headers()
        backend_url = 'unix:' + self.server.backend_address
        proxy_url = '{}:{}'.format(self.server.server_name,
                                   self.server.server_port)
        info = INFO.format(version=VERSION, proxy_url=proxy_url,
                           backend_url=backend_url,
                           connected=self.server.conn.is_connected())
        self.wfile.write(info.encode('utf-8'))

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Content-type", "text/plain")
        self.addCORS()
        self.end_headers()

    def do_POST(self):
        request_length = int(self.headers['Content-Length'])
        request_content = self.rfile.read(request_length)
        # self.log_message("Headers:  {}".format(self.headers))
        # self.log_message("Request:  {}".format(request_content))

        try:
            response_content = self.server.process(request_content)
            # self.log_message("Response: {}".format(response_content))

            self.send_response(200)
            self.send_header("Content-type", "application/json")
            self.send_header("Content-length", len(response_content))
            self.addCORS()
            self.end_headers()
            self.wfile.write(response_content)
        except BackendError as err:
            self.send_response(502)
            error_msg = str(err).encode('utf-8')
            # TODO: Send as JSON-RPC response
            self.send_header("Content-type", "text/plain")
            self.send_header("Content-length", len(error_msg))
            self.end_headers()
            self.wfile.write(error_msg)
            self.log_message("Backend Error: {}".format(err))

        # TODO: Handle other exceptions as error 500.

    def addCORS(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "content-type")


class Proxy(HTTPServer):

    def __init__(self, proxy_url, backend_path):
        self.proxy_url = proxy_url
        url = urlparse(proxy_url)
        assert url.scheme == 'http'
        proxy_address = url.hostname, url.port

        super(Proxy, self).__init__(proxy_address, HTTPRequestHandler)

        self.backend_address = path.expanduser(backend_path)

    def process(self, request):
        self.conn.sendall(request)

        response = b''
        while True:
            r = self.conn.recv(BUFSIZE)
            if not r:
                break
            if r[-1] == DELIMITER:
                response += r[:-1]
                break
            response += r

        return response

    def run(self):
        self.conn = get_ipc_connector(self.backend_address)
        self.conn.check_connection(timeout=BACKEND_CONNECTION_TIMEOUT)

        print("Dopple JSON-RPC HTTP Proxy: {} -> {}".format(
            self.backend_address, self.proxy_url), file=sys.stderr, flush=True)
        self.serve_forever()


if sys.platform == 'win32':
    DEFAULT_BACKEND_PATH = r'\\.\pipe\geth.ipc'
    BACKEND_PATH_HELP = "Named Pipe of a backend RPC server"
else:
    DEFAULT_BACKEND_PATH = '~/.ethereum/geth.ipc'
    BACKEND_PATH_HELP = "Unix Socket of a backend RPC server"

DEFAULT_PROXY_URL = 'http://127.0.0.1:8545'
PROXY_URL_HELP = "URL for this proxy server"


def parse_args():
    parser = ArgumentParser(
        description='Dopple HTTP Proxy for JSON-RPC servers',
        formatter_class=ArgumentDefaultsHelpFormatter
    )

    parser.add_argument('backend_path', nargs='?',
                        default=DEFAULT_BACKEND_PATH,
                        help=BACKEND_PATH_HELP)
    parser.add_argument('proxy_url', nargs='?',
                        default=DEFAULT_PROXY_URL,
                        help=PROXY_URL_HELP)
    return parser.parse_args()


def run(proxy_url=DEFAULT_PROXY_URL, backend_path=DEFAULT_BACKEND_PATH):
    proxy = Proxy(proxy_url, backend_path)
    try:
        proxy.run()
    except KeyboardInterrupt:
        proxy.shutdown()


def run_daemon(proxy_url=DEFAULT_PROXY_URL, backend_path=DEFAULT_BACKEND_PATH):
    proxy = Proxy(proxy_url, backend_path)
    th = threading.Thread(name='dopple', target=proxy.run)
    th.daemon = True
    th.start()
    return proxy

def main():
    args = parse_args()
    run(args.proxy_url, args.backend_path)

if __name__ == '__main__':
    main()

